src/main/java/io/jenkins/plugins/coverage/adapter/DListingReportAdapter.java

```diff
@@ -0,0 +1,195 @@
+package io.jenkins.plugins.coverage.adapter;
+
+import com.google.common.collect.Lists;
+import hudson.Extension;
+import io.jenkins.plugins.coverage.adapter.parser.CoverageParser;
+import io.jenkins.plugins.coverage.exception.CoverageException;
+import io.jenkins.plugins.coverage.targets.CoverageElement;
+import io.jenkins.plugins.coverage.targets.CoverageResult;
+import org.jenkinsci.Symbol;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+
+import javax.xml.XMLConstants;
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+
+import javax.annotation.Nonnull;
+import javax.xml.parsers.ParserConfigurationException;
+import java.io.File;
+import java.util.List;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.stream.Collectors;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public final class DListingReportAdapter extends CoverageReportAdapter {
+
+    private static final String UNCOVERED = "0000000";
+    private static final String NON_EXECUTABLE_CODE = "";
+
+    @DataBoundConstructor
+    public DListingReportAdapter(String path) {
+
+        super(path == null || path.equals("") ? "*.lst" : path);
+        setMergeToOneReport(true);
+    }
+
+    @Override
+    public Document convert(File source) throws CoverageException {
+        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
+        try {
+            factory.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
+        } catch (ParserConfigurationException e) {
+            e.printStackTrace();
+        }
+
+        factory.setNamespaceAware(true);
+        DocumentBuilder builder;
+        try {
+            builder = factory.newDocumentBuilder();
+        } catch (ParserConfigurationException e) {
+            throw new CoverageException(e);
+        }
+
+        Document document = builder.newDocument();
+        Element rootElement = document.createElement("report");
+        rootElement.setAttribute("name", "D Listing Coverage");
+        document.appendChild(rootElement);
+
+        Element fileElement = document.createElement("file");
+
+        try (BufferedReader br = Files.newBufferedReader(Paths.get(source.getPath()))) {
+            List<String> list = br.lines().collect(Collectors.toList());
+
+            if (list.size() == 0) {
+                return document;
+            }
+
+            String lastLine = list.get(list.size() - 1);
+            String moduleFileName = extractModuleFileName(lastLine);
+
+            if (moduleFileName == null) {
```

Comment:
Maybe we need to throw an exception instead of return an empty report?

Reply:
yes, will change